File: fim_api/python/lifecycle.py
from fog05 import FIMAPI
from fog05sdk.interfaces.FDU import FDU
import uuid
import json
import sys
import os


def read_file(filepath):
    with open(filepath, 'r') as f:
        data = f.read()
    return data


def main(ip, fdufile, netfile):
    a = FIMAPI(ip)

    nodes = a.node.list()
    if len(nodes) == 0:
        print('No nodes')
        exit(-1)

    print('Nodes:')
    for n in nodes:
        print('UUID: {}'.format(n))

    fdu_d = FDU(json.loads(read_file(fdufile)))
    net_d = json.loads(read_file(netfile))

    n_uuid = net_d.get('uuid')




    #n1 = 'a2d358aa-af2b-42cb-8d23-a89e88b97e5c' #fosmed
    n1 = '4a560914-1c3e-4966-9fa8-7f0acc903253' #nuc
    # n1 = '53712df2-9649-4a21-be2e-80eed00ff9ce' #ubuntuvm1
    #n1 = 'a4589fae-0493-40cf-b976-2d03020d060d' #foskvm
    # n1 = '53712df2-9649-4a21-be2e-80eed00ff9ce' #ubuntuvm1
    # n1 = 'd07b095f-7948-4f9b-95cc-c61029f6c3c3' #fosdbg

    input("Press enter to create network")
    a.network.add_network(net_d)
    a.network.add_network_to_node(net_d, n1)

    input('press enter to onboard descriptor')
    res = a.fdu.onboard(fdu_d)
    e_uuid = res.get_uuid()
    input('Press enter to intantiate')
    inst_info = a.fdu.instantiate(e_uuid, n1)
    instid = inst_info.get_uuid()


    # input('Press enter to stop')
    # a.entity.stop(e_uuid, n1, i_uuid)
    # input('Press enter to clean')
    # a.entity.clean(e_uuid, n1, i_uuid)
    # input('Press enter to undefine')
    # a.entity.undefine(e_uuid, n1)

    # input('Press enter to migrate')
    input('Press get info')
    info = a.fdu.instance_info(instid)
    print(info.to_json())

    #res = a.entity.migrate(e_uuid, i_uuid, n1, n2)
    #print('Res is: {}'.format(res))
    input('Press enter to terminate')

    a.fdu.terminate(instid)
    a.fdu.offload(e_uuid)
    input("Press enter to remove network")
    a.network.remove_network_from_node(n_uuid, n1)
    a.network.remove_network(n_uuid)

    exit(0)


if __name__ == '__main__':
    if len(sys.argv) < 4:
        print('[Usage] {} <yaks ip:port> <path to fdu descripto> <path to net descriptor>'.format(
            sys.argv[0]))
        exit(0)
    main(sys.argv[1], sys.argv[2], sys.argv[3])

File: fim_api/go/src/lifecycle.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"

	"github.com/eclipse-fog05/api-go/fog05"
	"github.com/eclipse-fog05/sdk-go/fog05sdk"
)

func check(e error) {
	if e != nil {
		panic(e)
	}

}

const n1 = "4a560914-1c3e-4966-9fa8-7f0acc903253" //nuc

// const n1 = "53712df2-9649-4a21-be2e-80eed00ff9ce" //ubuntuvm1local

func main() {
	args := os.Args[1:]
	api, err := fog05.NewFIMAPI(args[0], nil, nil)
	check(err)

	fmt.Printf("Nodes:\n")
	nodes, err := api.Node.List()
	check(err)

	for _, n := range nodes {
		fmt.Printf("UUID: %s\n", n)
	}

	data, err := ioutil.ReadFile(args[1])
	check(err)

	var fduDescriptor fog05sdk.FDU
	json.Unmarshal(data, &fduDescriptor)

	data, err = ioutil.ReadFile(args[2])
	check(err)

	var vNetDescriptor fog05sdk.VirtualNetwork
	json.Unmarshal(data, &vNetDescriptor)

	fmt.Printf("Press enter to create network\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')

	err = api.Network.AddNetwork(vNetDescriptor)
	check(err)
	vNetR, err := api.Network.AddNetworkToNode(n1, vNetDescriptor)
	check(err)

	fmt.Printf("Result of Network creation\n%+v\n", vNetR)

	fmt.Printf("Press enter to onboard descriptor\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(5 * time.Second)

	fduD, err := api.FDU.Onboard(fduDescriptor)
	check(err)
	fmt.Printf("Result of onboarding\n%+v\n", fduD)

	fID := *fduD.UUID

	fmt.Printf("Press enter to define\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')

	// time.Sleep(5 * time.Second)

	fduR, err := api.FDU.Define(n1, fID)
	check(err)
	fmt.Printf("Result of define:\n%+v\n", fduR)

	iID := fduR.UUID

	fmt.Printf("Press enter to configure\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(30 * time.Second)
	_, err = api.FDU.Configure(iID)
	check(err)

	fmt.Printf("Press enter to start\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(30 * time.Second)
	_, err = api.FDU.Start(iID)
	check(err)

	fmt.Printf("Press enter to get instance info\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(30 * time.Second)
	info, err := api.FDU.InstanceInfo(iID)
	check(err)

	fmt.Printf("Instance Info:\n%+v\n", info)

	fmt.Printf("Press enter to stop\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(30 * time.Second)
	_, err = api.FDU.Stop(iID)
	check(err)

	fmt.Printf("Press enter to Clean\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(30 * time.Second)
	_, err = api.FDU.Clean(iID)
	check(err)

	fmt.Printf("Press enter to Remove\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
	// time.Sleep(30 * time.Second)
	_, err = api.FDU.Undefine(iID)
	check(err)

	_, err = api.FDU.Offload(fID)
	check(err)

	fmt.Printf("Press enter to Remove Network\n")
	bufio.NewReader(os.Stdin).ReadBytes('\n')

	_, err = api.Network.RemoveNetworkFromNode(n1, vNetR.UUID)
	check(err)
	err = api.Network.RemoveNetwork(vNetR.UUID)
	check(err)

}
